### PR TITLE
Fixed integer over flow bug 

### PR DESCRIPTION
**Pull request by lessthanoptimal**
- Fixed integer over flow bug when defining sparse matrices and creating a random sparse matrix
- Code formatting

### CODE REVIEW

File: main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java [hunk 1]
```diff
@@ -45,25 +45,30 @@ public boolean isStructureValid(DMatrixSparse m) {
         return CommonOps_DSCC.checkStructure((DMatrixSparseCSC)m);
     }
 
+    @Test
+    public void constructor_tooLarge() {
+        assertThrows(IllegalArgumentException.class,()->new DMatrixSparseCSC(1_000_000_000,1_000_000_000,4));
+    }
+
```
Comment (FlorentinD) [inline main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java]: `10^9` is actually smaller than `Integer.MAX_VALUE`. So this would be actually a valid sparse matrices.

File: main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java [hunk 2]
```diff
@@ -45,18 +44,21 @@
      * @param rand Random number generator
      * @return Randomly generated matrix
      */
-    public static DMatrixSparseCSC rectangle(int numRows , int numCols , int nz_total ,
-                                             double min , double max , Random rand ) {
+    public static DMatrixSparseCSC rectangle( int numRows, int numCols, int nz_total,
+                                              double min, double max, Random rand ) {
+
+        if (UtilEjml.exceedsMaxMatrixSize(numRows, numCols))
+            throw new IllegalArgumentException("You have specified a matrix which exceeds the size of an integer");
```
Comment (FlorentinD) [inline main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java]: this more like for temporary fix for the random sparse matrix right?
Maybe we also want to assert, that `nz_total` has a positive value?

File: main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java [hunk 3]
```diff
@@ -90,42 +91,42 @@ public static DMatrixSparseCSC rectangle(int numRows , int numCols , int nz_tota
...
      * @return Randomly generated matrix
      */
-    public static DMatrixSparseCSC symmetric( int N , int nz_total ,
-                                              double min , double max , Random rand) {
+    public static DMatrixSparseCSC symmetric( int N, int nz_total,
+                                              double min, double max, Random rand ) {
 
         // compute the number of elements in the triangle, including diagonal
-        int Ntriagle = (N*N+N)/2;
+        int Ntriagle = (N*N + N)/2;
```
Comment (FlorentinD) [inline main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java:102]: we also need to check here for an overflow

File: main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java [hunk 4]
```diff
@@ -235,13 +237,13 @@ public static int nonzero( int numRows , int numCols , double minFill , double m
      * @param rand random number generator
      * @return Random matrix
      */
-    public static DMatrixSparseCSC triangle( boolean upper , int N , double minFill , double maxFill , Random rand ) {
-        int nz = (int)(((N-1)*(N-1)/2)*(rand.nextDouble()*(maxFill-minFill)+minFill))+N;
+    public static DMatrixSparseCSC triangle( boolean upper, int N, double minFill, double maxFill, Random rand ) {
+        int nz = (int)(((N - 1)*(N - 1)/2)*(rand.nextDouble()*(maxFill - minFill) + minFill)) + N;
```
Comment (FlorentinD) [inline main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java]: here again `(N-1)*(N-1)` could overflow

File: main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java (hunk 2 above)
Comment (FlorentinD) [reply main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java]: So maybe we could change the error message to something like `This method cannot create random sparse matrices, where the dense version would not fit`

File: main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java (hunk 1 above)
Comment (FlorentinD) [reply main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java]: (see other comments)

File: main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java [hunk 5]
```diff
@@ -45,25 +45,30 @@ public boolean isStructureValid(DMatrixSparse m) {
...
     @Test
     public void constructor_veryLarge() {
-        DMatrixSparseCSC a = new DMatrixSparseCSC(1_000_000_000,100_000_000,4);
+        DMatrixSparseCSC a = new DMatrixSparseCSC(1_000_000,1_000,4);
```
Comment (FlorentinD) [inline main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java]: I think this behaviour should still be allowed, as its not wrong to my point of view

File: main/ejml-core/src/org/ejml/data/DMatrixSparseCSC.java [hunk 6]
```diff
@@ -92,15 +89,17 @@ public DMatrixSparseCSC(int numRows , int numCols ) {
      * @param numCols Number of columns
      * @param arrayLength Initial maximum number of non-zero elements that can be in the matrix
      */
-    public DMatrixSparseCSC(int numRows , int numCols , int arrayLength) {
+    public DMatrixSparseCSC( int numRows, int numCols, int arrayLength ) {
+        if (UtilEjml.exceedsMaxMatrixSize(numRows, numCols))
+            throw new IllegalArgumentException("You have specified a matrix which exceeds the size of an integer");
```
Comment (FlorentinD) [inline main/ejml-core/src/org/ejml/data/DMatrixSparseCSC.java]: For sparse matrices this is actually valid to do so. The important point is, that the nz-values do not exceed Integer.MAX_VALUE

File: main/ejml-core/src/org/ejml/UtilEjml.java [hunk 7]
```diff
@@ -574,35 +577,44 @@ public static String fancyString(double value, DecimalFormat format, boolean has
...
     }
 
+    /**
+     * Checks to see if a matrix of this size will exceed the maximum possible value an integer can store, which is
+     * the max possible array size in Java.
+     */
+    public static boolean exceedsMaxMatrixSize( int numRows, int numCols ) {
+        if (numRows == 0 || numCols == 0)
+            return false;
+        return numCols > Integer.MAX_VALUE/numRows;
```
Comment (FlorentinD) [inline main/ejml-core/src/org/ejml/UtilEjml.java:618]: As mentioned in other comments, I think this check is valid for the current random matrix creation.
However in general a sparse random matrix can have at maximum `numRows == Integer.MAX_VALUE` and `numCols == Integer.MAX_VALUE`. The important thing is, that the nz-values do not exceed this limit.

**Review by FlorentinD — CHANGES_REQUESTED** [review]
TLDR: I think the check should not be applied in general for `DMatrixSparseCSC` but just for the current random matrix generator.

File: main/ejml-core/src/org/ejml/UtilEjml.java (hunk 7 above)
Comment (lessthanoptimal) [reply main/ejml-core/src/org/ejml/UtilEjml.java:618]: Isn't that what this is checking? Mathematically this is identical to `numCols*numRows > Integer.MAX_VALUE` but takes in account the possibility of `numCols*numRows` overflowing.

File: main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java (hunk 5 above)
Comment (lessthanoptimal) [reply main/ejml-core/test/org/ejml/data/TestDMatrixSparseCSC.java:55]: `nz_rows` can only take on values from 0 to `Integer.MAX_VALUE` if `numRows*numCols` exceeds that value then not every element can be referenced. The long term fix is to change it from int to long, pun intended.

File: main/ejml-core/src/org/ejml/UtilEjml.java (hunk 7 above)
Comment (lessthanoptimal) [reply main/ejml-core/src/org/ejml/UtilEjml.java:618]: Never mind I had a brain fart. Fixing all of this.

**Review by FlorentinD — APPROVED** [review]
LGTM :+1: 

As previously commented, in the other random-matrix dscc functions like `symmeteric` the check could be also added. But the original bug is resolved this way.

**Comment by lessthanoptimal** [comment]
Added additional checks and also noticed the unit tests were lacking so I improved a couple of those.
